src/web/controller.py

```diff
@@ -0,0 +1,184 @@
+"""Controller for the web app."""
+import base64
+import io
+import tempfile
+from typing import Dict, Optional, Tuple, Union
+
+from web.messages.requests import (
+    IndexBuiltInScenario,
+    RenderLayoutRequest,
+    SimulateRequest,
+)
+
+from neurotechdevkit.results import SteadyStateResult2D, SteadyStateResult3D
+from neurotechdevkit.scenarios import Scenario2D, Scenario3D
+from neurotechdevkit.scenarios.built_in import BUILT_IN_SCENARIOS
+
+
+class BuiltInScenariosShelf(object):
+    """Singleton class for storing the already instantiated built-in scenarios."""
+
+    scenarios: Dict[str, Scenario2D] = {}
+
+    def __new__(cls):
+        """Create a new instance of the BuiltInScenariosShelf class."""
+        if not hasattr(cls, "instance"):
+            cls.instance = super(BuiltInScenariosShelf, cls).__new__(cls)
+        return cls.instance
+
+    def get(self, scenario_id: str) -> Scenario2D:
+        """
+        Return the instantiated built-in scenario with the given id.
+
+        Args:
+            scenario_id: The id of the scenario.
+
+        Returns:
+            The built-in scenario.
+        """
+        if scenario_id not in self.scenarios:
+            builtin_scenario = BUILT_IN_SCENARIOS[scenario_id]()
+            builtin_scenario.make_grid()
+            builtin_scenario.compile_problem()
+            self.scenarios[scenario_id] = builtin_scenario
+        return self.scenarios[scenario_id]
```

Comment:
Not sure if this is the right place for this comment:

The scenarios currently error out when trying to render/run a center frequency other than `5e6`. The error is:
```
  File "/Users/charles/Development/clean-neurotechdevkit/src/web/controller.py", line 139, in get_simulation_image
    scenario.compile_problem()
  File "/Users/charles/Development/clean-neurotechdevkit/src/neurotechdevkit/scenarios/_base.py", line 341, in compile_problem
    assert self.grid is not None
           ^^^^^^^^^
  File "/Users/charles/Development/clean-neurotechdevkit/src/neurotechdevkit/scenarios/_base.py", line 299, in grid
    assert hasattr(self, "_grid")
AssertionError
```

Perhaps the scenarios are initialized here with `5e6`, then need to be re-initialized properly for other frequencies?

https://www.loom.com/share/d052365dd6a948de8693d379109799fd
https://www.loom.com/share/4ecc435c13ab47199c20976a2fa35b39

(As a side note, this screen-recording shows that the error message appears when I change the material color, which triggers a `render_layout`)